File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/1. Learn Bit Manipulation/javascript/07.Divide two numbers using bit maipulation.js
/**
QUESTION:
Divide two integers `dividend` and `divisor` without using multiplication, division, or modulus.

Approach:
Work with absolute values and repeatedly subtract the largest shifted divisor from the dividend, building the quotient bit by bit. Track the sign separately.

Time Complexity: O(log n)
Space Complexity: O(1)
*/

function divideTwoInteger(dividend, divisor) {
    if (divisor === 0) throw new Error("Division by zero");
    if (dividend === divisor) return 1;

    const isNegative = (dividend < 0 && divisor >= 0) || (dividend >= 0 && divisor < 0);
    let a = Math.abs(dividend);
    const b = Math.abs(divisor);
    let ans = 0;

    while (a >= b) {
        let q = 1;
        while (a > (b << q)) q++;
        ans += (1 << (q - 1));
        a -= (b << (q - 1));
    }

    return isNegative ? -ans : ans;
}

module.exports = { divideTwoInteger };

File: Strivers-A2Z-DSA-Sheet/06.Bit Manipulation/1. Learn Bit Manipulation/python/07.Divide two numbers using bit maipulation.py
"""QUESTION:
Divide two integers without using multiplication, division, or modulus.

Approach:
Work with absolute values. Repeatedly subtract the largest shifted divisor that fits inside the dividend, accumulating the quotient via powers of two. Apply the overall sign at the end.

Time Complexity: O(log n)
Space Complexity: O(1)
"""

def divide_two_integer(dividend: int, divisor: int) -> int:
    if divisor == 0:
        raise ZeroDivisionError("Division by zero")
    if dividend == divisor:
        return 1

    is_negative = (dividend < 0) ^ (divisor < 0)
    a = abs(dividend)
    b = abs(divisor)
    ans = 0

    while a >= b:
        q = 1
        while a > (b << q):
            q += 1
        ans += 1 << (q - 1)
        a -= b << (q - 1)

    return -ans if is_negative else ans
